[WinEH] Fix try scopes leaking to caller on inline

MuellerMP opened the pull request:
This fixes issue #164169

When inlining functions compiled with -EHa, try scope terminators might need to be inserted before inlined returns. This prevents leaking try scopes over to the caller. Try scopes can be ended before a ret due to the unwinder forwarding exceptions in the seh epilog to the caller.

MuellerMP commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4128,47 +4107,78 @@ void CodeGenFunction::EmitFunctionEpilog(
+                              RetDbgLoc);
+
+  if (SehTryEndInvokeDest && isReturnReachable(ReturnBlock))
+    EmitSehTryScopeEnd(SehTryEndInvokeDest);
```
I think it is possible to formulate this as a `NormalCleanup` on the EHStack (similar how SEH finally scopes are pushed). I initially put this here (and fixed up WinEHPrepare to work with already ended scopes) so i can emit the scope end just before the ret and keep return value loads inside the try scope.
But I do agree that using the EHStack is probably the better approach here.

MuellerMP committed:
[WinEH] Fix try scopes leaking to caller on inline

This fixes issue #164169

When inlining functions compiled with -EHa, try scope terminators might need to be inserted before inlined returns.
This prevents leaking try scopes over to the caller.
Try scopes can be ended before a ret due to the unwinder forwarding exceptions in the seh epilog to the caller.

efriedma-quic reviewed (commented):
This seems like basically the right approach.

MuellerMP commented on llvm/lib/CodeGen/WinEHPrepare.cpp:
```diff
@@ -341,9 +341,11 @@ void llvm::calculateSEHStateForAsynchEH(const BasicBlock *BB, int State,
         State = EHInfo.InvokeStateMap[cast<InvokeInst>(TI)];
       else if (Fn && Fn->isIntrinsic() &&
-               Fn->getIntrinsicID() == Intrinsic::seh_try_end)
+               Fn->getIntrinsicID() == Intrinsic::seh_try_end) {
```
This reuses the invoke state map for try.end in a fashion already available for the cxx state numbering algorithm.
It allows to terminate multiple SEH scopes at once without needing an additional cleanup scopes per nested try.
It works by looking up the state for the matching unwind destination and then decrements the state number (which is expressed by the ToState assignment).

I added a nested try catch test which showcases this (without this patch we will end up in `Exit` instead of having the exception propagate to the caller).

MuellerMP replied:
Updated my patch: the current patch now reverts back to emitting a try end for every eha try. The related invoke destination is now also present in the TerminateTryScope cleanup struct.

efriedma-quic commented on llvm/test/CodeGen/WinEH/wineh-statenumbering-seh-try-end.ll:
```diff
@@ -0,0 +1,170 @@
+
+@p = dso_local global ptr null, align 8
+
+define dso_local noundef i32 @main(i32 noundef %argc, ptr noundef readnone captures(none) %argv) local_unnamed_addr personality ptr @__C_specific_handler {
```
If this was originally C++, maybe put the corresponding C++ code for reference.

MuellerMP replied:
test is now obsolete with the new approach

efriedma-quic reviewed (commented):
Not sure when @rnk will have a chance to respond; let's try to move forward without him.

MuellerMP commented on clang/lib/CodeGen/CGException.cpp:
```diff
@@ -1675,6 +1701,8 @@ void CodeGenFunction::EmitSEHTryStmt(const SEHTryStmt &S) {
       if (SEHTryEpilogueStack.size() == 1) // outermost only
         TryBB = Builder.GetInsertBlock();
+      if (!SehTryEndInvokeDest)
+        SehTryEndInvokeDest = getInvokeDest();
```
Update: My current solution is now to just create the unwind destination basicblock early fo the TerminateTryScope instances which i can cache with `setCachedEHDispatchBlock`.

efriedma-quic commented on clang/lib/CodeGen/CGException.cpp:
```diff
@@ -635,14 +635,30 @@ void CodeGenFunction::EmitCXXTryStmt(const CXXTryStmt &S) {
+  unsigned NumHandlerInfos = NumHandlers > 0 ? NumHandlers - 1 : 0;
+  llvm::BasicBlock *DispatchBlock = nullptr;
+  llvm::SmallVector<HandlerInfo> HandlerInfos;
+  HandlerInfos.reserve(NumHandlerInfos);
```
It's not worth the extra code here to try to reserve the exact amount in the array.

efriedma-quic commented on clang/test/CodeGen/windows-seh-EHa-Inline1.cpp:
```diff
@@ -0,0 +1,62 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 6
+// RUN: %clang_cc1 -O3 -triple x86_64-windows -fasync-exceptions -fcxx-exceptions -fexceptions -fms-extensions -x c++ -Wno-implicit-function-declaration -emit-llvm %s -o - | FileCheck %s
```
We generally prefer clang tests to run with optimizations disabled: it makes it easier to see the correspondence between what we're testing and what's actually getting generated, and it makes the tests less sensitive to optimizer changes.  (There's a minor downside that it doesn't always capture the end-to-end result, but usually a set of clang tests plus a set of LLVM optimizer tests is sufficient.)

efriedma-quic reviewed (commented):
Does this fix #62723?

MuellerMP commented:
Okay so I just realized how I can reduce this fix to 32 lines basically and resolve the issues I had with WinEHPrepare.
`PopCleanupBlock` already provides the mechanism for try scope termination if we have a cleanup scope for e.g. a destructor call.
If we do not have any cleanup scope available the scope termination mechanism is never called though.
So the best fix I can think of now is to just push a noop `TerminateTryScope` cleanup on the EHStack.

Since we can push this even after the Catch this resolved my issues with WinEHPrepare, since the state handling algorithm now also traverses the scope end in the correct order.
So I deleted my backend test and fixes in WinEhPrepare and fixed up my frontend tests to be executed without opt passes.

This also resolves the FIXME in the windows-seh-EHa-CppCatchReturn.cpp test.

rnk reviewed (commented):
Thanks for following up. I had some minor nits to fix, but otherwise it looks good.

MuellerMP committed:
Update comment

Co-authored-by: Reid Kleckner <rkleckner@nvidia.com>

MuellerMP commented:
ping

GkvJwa commented on clang/lib/CodeGen/CGException.cpp:
```diff
@@ -666,8 +667,11 @@ void CodeGenFunction::EnterCXXTryStmt(const CXXTryStmt &S, bool IsFnTryBlock) {
+      if (getLangOpts().EHAsynch) {
         EmitSehTryScopeBegin();
+        // Push cleanup to emit the end of the scope
+        EHStack.pushCleanup<TerminateTryScope>(NormalCleanup);
```
Add SEHTryEpilogueStack.push_back(EHStack.stable_begin()); better?
Then PopCleanupBlock(); in CodeGenFunction::ExitSEHTryStmt

like:
```
void CodeGenFunction::EnterSEHTryStmt(const SEHTryStmt &S) {             
    // ... existing code ...
    if (getLangOpts().EHAsynch) {                                          
      EHStack.pushCleanup<TerminateTryScope>(NormalCleanup);               
      SEHTryEpilogueStack.push_back(EHStack.stable_begin()); // new member 
    }                                                                      
  }                                                                        
                                                                           
  void CodeGenFunction::ExitSEHTryStmt(const SEHTryStmt &S) {              
    if (S.getFinallyHandler()) { PopCleanupBlock(); return; }
                                                                           
    if (getLangOpts().EHAsynch) {                                          
      assert(!SEHTryEpilogueStack.empty());                                
      auto top = SEHTryEpilogueStack.pop_back_val();                       
      assert(EHStack.stable_begin() == top &&                              
             "TerminateTryScope cleanup got reordered");
      PopCleanupBlock();   // guaranteed to pop the one we pushed          
    }                                                                      
    // ... existing code ...                                               
  }
```

or, similar to my approach, add a marker
```
EHStack.pushCleanup<TerminateTryScope>(
      static_cast<CleanupKind>(NormalCleanup | SEHTryScopeCleanup));

if (Scope.isSEHTryScopeCleanup() || Scope.isSEHFinallyCleanup())
    EmitSehTryScopeEnd();
  else
    EmitSehCppScopeEnd();
```

EDIT: Use ExitCXXTryStmt to determine if the top of EHStack has a relatively loose Cleanup
If the changes are minor, you can do it like this
```
static bool hasCatchAllHandler(const CXXTryStmt &S) {
    for (unsigned I = 0, E = S.getNumHandlers(); I != E; ++I)
      if (!S.getHandler(I)->getExceptionDecl())
        return true;
    return false;
  }

void CodeGenFunction::ExitCXXTryStmt(const CXXTryStmt &S, bool
  IsFnTryBlock) {
    bool HasEHaTryTerminator =
        getLangOpts().EHAsynch && hasCatchAllHandler(S);

    if (HasEHaTryTerminator) {
      PopCleanupBlock();
    }
```
